Quickstart Things

egekorkan opened the pull request:
fixes https://github.com/eclipse-thingweb/node-wot/issues/1063

and also addresses https://github.com/w3c/wot-cg/pull/67

It is not done yet since:

- [x] I have TS errors on the coffee machine
- [x] I need to implement intervals on the smart clock
- [x] I need to test the behavior
- [x]  I need to understand how to fix the resolve type error (see comment below)

egekorkan commented:
@relu91 and @danielpeintner I am aware of the failing eslint but I think I am not understanding how to work with the InteractionOutput type. In the simple coffee machine, I want to return a Promise since the action takes a while. I did not manage to avoid the error by trying different combinations. I have put the ts-ignore statements in order to test the behavior. 

Other than this: 
- The behavior of the devices are as intended. They are super simple to test, which was one of the goals of this implementation.
- I did not copy them over to root level examples folder yet. Before I do that, I will change the sensor to MQTT and the clock to CoAP.

danielpeintner commented:
> @relu91 and @danielpeintner I am aware of the failing eslint but I think I am not understanding how to work with the InteractionOutput type. In the simple coffee machine, I want to return a Promise since the action takes a while. I did not manage to avoid the error by trying different combinations. I have put the ts-ignore statements in order to test the behavior.

I removed ts-ignore statements and pushed an update (see https://github.com/eclipse-thingweb/node-wot/pull/1064/commits/511602d0e7ac086fa4b2706d48071b56989a22cf). Please let me know whether it works as expected.

BTW, there are several log statements that need to be removed.. I think
* `console.log("here");`
* `console.log("here2");`
* `console.log("here3");`
* `console.log("here4");`

danielpeintner commented on packages/examples/src/quickstart/simple-coffee-machine.ts:
```diff
@@ -0,0 +1,136 @@
+                if (coffeeType === "espresso") {
+                    console.log("here");
+                    if (waterAmount <= 10 || beansAmount <= 10) {
+                        console.log("here4");
```
remove?

danielpeintner reviewed (changes requested):
Please remove unnecessary log statements

egekorkan commented:
@danielpeintner your contributions fixed the issues. I did:
- Changed clock to CoAP
- I have noticed an old example in mqtt binding readme and fixed it
- There is are weird problems with MQTT:
  - The import has an error but I have added an ignore statement.
  - There is no connection to the broker. The TD is not logged and events are not emitted. The example in the README works just fine. I feel like I am not seeing something...

danielpeintner commented on packages/binding-mqtt/README.md:
```diff
@@ -31,7 +31,7 @@ MqttBrokerServer = require("@node-wot/binding-mqtt").MqttBrokerServer;
 // create Servient add MQTT binding
 let servient = new Servient();
-servient.addServer(new MqttBrokerServer({uri:"mqtt://test.mosquitto.org"}));
+servient.addServer(new MqttBrokerServer({ uri: "mqtt://test.mosquitto.org" }));
```
I don't think this change belongs to this PR ... 
It is not a big deal if we leave it in but somewhat confusing

egekorkan replied:
I can prepare another PR for it. I was using it for testing :)

danielpeintner commented:
> There is are weird problems with MQTT:
> 
>     * The import has an error but I have added an ignore statement.

We should keep track of these ignore statements. We used to have https://github.com/eclipse-thingweb/node-wot/issues/586 for it.
Anyhow, we can open another issue otherwise it gets forgotten and I think we should look into "what" causes the issue

egekorkan commented:
@danielpeintner the MQTT example was working the whole time, I had changed the file name and was running the old js file the whole time... 🤦

egekorkan commented:
Given that the tests are also passing, I would like to merge this asap and deploy it to have it ready for TPAC. @danielpeintner @relu91 any further changes/review?

danielpeintner commented on packages/examples/src/quickstart/smart-clock.ts:
```diff
@@ -0,0 +1,96 @@
+// This is an example Thing which is a smart clock that runs 60 times faster than real time, where 1 hour happens in 1 minute.
+
+import { Servient } from "@node-wot/core";
+// import { HttpServer } from "@node-wot/binding-http";
```
remove? or have the same commented line for 

servient.addServer(new HttpServer(...));

danielpeintner commented on packages/examples/src/quickstart/smart-clock.ts:
```diff
@@ -0,0 +1,96 @@
+                observable: true,
+                type: "object",
+                properties: {
+                    minute: {
+                        type: "integer",
+                        minimum: 0,
+                        maximum: 59,
+                    },
+                    hour: {
+                        type: "integer",
+                        minimum: 0,
+                        maximum: 23,
+                    },
```
There are leap seconds and we would need to have 60 or 24 respectively but I guess this becomes too complex :-)

egekorkan replied:
I wasn't aware of this :D It would indeed complicate things, I want to keep these Things as simple as possible

danielpeintner reviewed (changes requested):
There are some things we can/should fix/integrate. Otherwise the PR looks good to me. Let's wait for @relu91

relu91 commented on packages/examples/README.md:
````diff
@@ -14,16 +14,7 @@ see https://github.com/eclipse-thingweb/node-wot/issues/171.
 ### Workflow
 
 1. Run `npm run build`
-2. Remove the following 3/4 lines in JS files of folder `dist/`
-
-```
-Object.defineProperty(exports, "__esModule", { value: true });
-require("wot-typescript-definitions");
-let WoT;
-let WoTHelpers;
-```
-
-3. Copy the according JS file(s) to
````
The `Object.defineProperty(exports, "__esModule", { value: true });` should be *always* generated (unless you are not importing anything). I'd keep the readme as it was and complete the PR by also copying the files to the dedicated js folder (for demo purposes using plain JS may be even more inclusive). 

Regarding lines 21-23, I'm 90% sure that they are generated only when the script is meant to be used as node-wot script, meaning that WoT and WoTHelpers variables are automatically injected by the CLI context. They are not present in the these quickstart examples because @egekorkan is using node-wot as a library dependency.

egekorkan replied:
So those lines are not generated when I run npm run build since WoT variables are injected by the CLI context. I have reverted these lines since we should restructure this folder and put instructions to each readme. Currently, this is in the root folder of the examples and these do not apply to all examples.

I am indeed running node-wot as a dependency in order to create 3 different servient that have different protocols. This way, one has a higher tendency to use node-wot to communicate with 3 Things using different protocols.

relu91 commented on packages/examples/src/quickstart/presence-sensor.ts:
```diff
@@ -0,0 +1,65 @@
+
+import { Servient } from "@node-wot/core";
+// eslint-disable-next-line workspaces/require-dependency
+import { MqttBrokerServer } from "@node-wot/binding-mqtt";
```
It was only a matter of ordering, fixed in https://github.com/eclipse-thingweb/node-wot/pull/1064/commits/0ba43cddd2c2d43cc6f4486f482f2daa14648e03 and https://github.com/eclipse-thingweb/node-wot/pull/1064/commits/cf0c6fd65a2de51ee0fa852a60af66027bf8ad65

egekorkan replied:
Thanks!

relu91 reviewed (changes requested):
As said in the comment above before merging we should copy the js files to the `examples` folder. I reviewed the code I think it is a good toy example, but in the next few days I'll try to use the Things lives and I might update it with fixes and improvements.

egekorkan commented on packages/examples/src/quickstart/simple-coffee-machine.ts:
```diff
@@ -0,0 +1,137 @@
+                properties: {
+                    water: {
+                        type: "integer",
+                        minimum: 10,
```
I was thinking of normal working conditions but true that someone can empty it manually :D Fixing :)

egekorkan commented:
@danielpeintner @relu91 I have implemented the feedback and copied over to the root examples folder.

relu91 commented on examples/quickstart/presence-sensor.js:
```diff
@@ -0,0 +1,60 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR W3C-20150513
+ ********************************************************************************/
+Object.defineProperty(exports, "__esModule", { value: true });
```
```suggestion

```

relu91 commented on examples/quickstart/simple-coffee-machine.js:
```diff
@@ -0,0 +1,130 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR W3C-20150513
+ ********************************************************************************/
+Object.defineProperty(exports, "__esModule", { value: true });
```
```suggestion

```

relu91 commented on examples/quickstart/smart-clock.js:
```diff
@@ -0,0 +1,87 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR W3C-20150513
+ ********************************************************************************/
+Object.defineProperty(exports, "__esModule", { value: true });
```
```suggestion

```

relu91 reviewed (changes requested):
You forgot to remove the first line, I added suggestions so that I can do a quick commit.
